### PR TITLE
Fix snapshotting symlinked directories

### PR DESCRIPTION
**Pull request by wolfs**
the name of the symlinked directory was wrong, as well as the absolute paths as soon as more of two symlinks have been involved.

### CODE REVIEW

File: subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java [hunk 1]
```diff
@@ -177,6 +197,46 @@ public boolean test(String element) {
         }
     }
 
+    private interface SymbolicLinkMappings {
+        String remapAbsolutePath(Path path);
+        @CheckReturnValue
+        SymbolicLinkMappings withNewMapping(String source, String target, RelativePathTracker currentRelativePathTracker);
+        Iterable<String> getRemappedSegments(Iterable<String> segments);
+    }
```
Comment (lptr) [inline subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java:205]: I don't think there's a point in having a `private` interface for a single `private` inner class.

File: subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java [hunk 2]
```diff
@@ -177,6 +197,46 @@ public boolean test(String element) {
...
+                .addAll(symbolicLinkMappings)
+                .build();
+            return new DefaultSymbolicLinkMappings(newMappings, getRemappedSegments(currentRelativePathTracker.getSegments()));
+        }
+
+        @Override
+        public Iterable<String> getRemappedSegments(Iterable<String> segments) {
+            return Iterables.concat(prefixRelativePath, segments);
+        }
+    }
```
Comment (lptr) [inline subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java:238]: I wonder if we could simplify this significantly. I think we only need a single type, something like:


```java
private static class SymbolicLinkMapping {

    public String remap(Path path) {
        return path.toString();
    }

    public SymbolicLinkMapping addMapping(String source, String target, RelativePathTracker currentRelativePathTracker) {
        SymbolicLinkMapping parent = this;
        return new SymbolicMapping() {
            public String remap(Path path) {
                String parentPath = parent.remap(path);
                return // actually remap path
            }
        }
    }
}
```

File: subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotterStatistics.java [hunk 3]
```diff
@@ -104,39 +104,39 @@ public String toString() {
     }
 
     abstract class CollectingFileVisitor implements FileVisitor<Path> {
-        private final Collector collector;
+        protected final Collector collector;
```
Comment (lptr) [inline subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotterStatistics.java:107]: Why does this need to be `protected`?

File: subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java (hunk 1 above)
Comment (wolfs) [reply subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java:205]: There are two implementations of the interface: The empty remapper and the other ones.

File: subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotterStatistics.java (hunk 3 above)
Comment (wolfs) [reply subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotterStatistics.java:107]: We create a new `PathVisitor`, and the new path visitor needs the collector. I can also make the collector a field in `PathVisitor` as well.

File: subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/MerkleDirectorySnapshotBuilder.java [hunk 4]
```diff
@@ -60,6 +60,10 @@ public void visitLeafElement(FileSystemLeafSnapshot snapshot) {
         collectEntry(snapshot);
     }
 
+    public void collectCompleteDirectory(CompleteDirectorySnapshot directorySnapshot) {
+        collectEntry(directorySnapshot);
+    }
```
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/MerkleDirectorySnapshotBuilder.java:65]: I renamed it to `visitDirectory`. Using `visitEntry` which shouldn't be called when using `enter/leave` is more confusing I think.

File: subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java (hunk 2 above)
Comment (wolfs) [reply subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java]: I simplified it. There is now only one class, and we also don't need to loop at all.

**Comment by wolfs** [comment]
@lptr PTAL!

**Pull request merged**